include/micm/cuda/util/cuda_util.cuh

```diff
@@ -30,5 +33,54 @@ namespace micm
 
     /// @brief Get the cuBLAS handle for the current device
     cublasHandle_t& GetCublasHandle();
+
+    /// @brief Define a functor for the cudaStream unique pointer deleter
+    struct CudaStreamDeleter
+    {
+      void operator()(cudaStream_t* cuda_stream) const
+      {
+        if (cuda_stream != nullptr)
+        {
+          CHECK_CUDA_ERROR(cudaStreamDestroy(*cuda_stream), "CUDA stream finalization failed");
+          delete cuda_stream;
+          cuda_stream = nullptr;
+        }
+      }
+    };
+
+    /// @brief Define the smart pointer type using the functor for the custom deleter
+    using CudaStreamPtr = std::unique_ptr<cudaStream_t, CudaStreamDeleter>;
+
+    /// @brief Singleton class to manage CUDA streams
+    class CudaStreamSingleton
+    {
+    public:
+
+      CudaStreamSingleton() = default;
+      
+      ~CudaStreamSingleton() = default;
```

Comment:
I think this function should be defined, but it should call the `CleanUp` function below. That way we don't have to remember to call cleanup. In fact, if we never need to manually call `CleanUp`, then we could put the contents of `CleanUp` into the destructor

Reply:
Originally I implemented it exactly what you suggested, but the `CudaStream` resource was something different. When putting the `cleanup` function in the destructor, the CUDA driver is already turned down before the destructor is called and it will lead to an error saying `driver is shutting down`. That is why I have to call the `cleanup` function separately as shown below to clean up the CUDA resource explicitly and avoid GPU memory leak.